packages/wallet/ui/src/util/SuggestChain.js

```diff
@@ -44,41 +41,38 @@ const makeChainInfo = (
 };
 
 /**
- * @param {string} networkConfig URL
+ * @param {string} networkConfigHref
  * @param {object} io
  * @param {typeof fetch} io.fetch
  * @param {import('@keplr-wallet/types').Keplr} io.keplr
  * @param {typeof Math.random} io.random
  * @param {string} [caption]
  */
 export async function suggestChain(
-  networkConfig,
+  networkConfigHref,
   { fetch, keplr, random },
   caption = undefined,
 ) {
-  console.log('suggestChain: fetch', networkConfig); // log net IO
-  const res = await fetch(networkConfig);
+  console.log('suggestChain: fetch', networkConfigHref); // log net IO
+  const url = new URL(networkConfigHref);
+  const res = await fetch(url);
   if (!res.ok) {
     throw Error(`Cannot fetch network: ${res.status}`);
   }
 
-  const { chainName: chainId, apiAddrs, rpcAddrs } = await res.json();
-  assert.equal(apiAddrs.length, rpcAddrs.length);
-  const index = Math.floor(random() * rpcAddrs.length);
+  /** @type {import('@agoric/internal').NetworkConfig} */
+  const networkConfig = await res.json();
```

Comment:
I make this sort of type annotation all the time when I'm prototyping, but it's unsound. I suppose it's just documenting an assumption we were already making, so it's not a regression. But we should add a shape guard or the like. Or use type `unknown` so tsc forces us to validate each assumption we make about this input.

Please make some XXX annotation to say "this is not best practice; don't copy this idiom blindly".

Reply:
I think this is the order of bad to good:
1. implicit `any` (what we have all over the place)
2. constrain `any` (what is here)
3. unknown (so you have to validate on use, but a pain)
4. type guard (which is why I put the TODO for that)

I'm having trouble thinking up what the unsound type exploit is here. If someone controls network config, there are much worse things they can do

Reply:
Agreed on the order. I guess I was reading too fast and missed the TODO.

> I'm having trouble thinking up what the unsound type exploit is here.

I don't know that this is exploitable. But the unsoundness adds a burden when reviewing; I have to keep track of all the places where `networkConfig` is used, knowing that it's really an external input and might not be of the declared type.

Thanks for the type guard.